### PR TITLE
Support obj name keys

### PR DESCRIPTION
**Pull request by conorbev**
Hi,

This is based off the discussion at: https://groups.google.com/forum/#!topic/jmxtrans/yVgxuL27K4g which was initially around being able to specify *all* MBeans in a query.

I initially looked at doing this by putting JexlNamingStrategy into BaseOutputWriter (and therefore GraphiteWriter which is my real use case).

However it wasn't so simple as that because JexlNamingStrategy operates over the Result object and the information I wanted to put into the key is part of the ObjectName (the domain) so I was unable to use JexlNamingStrategy as it was.

While I'm OK with the idea of JexlNamingStrategy and think it could be made to work for this case after the changes I've made, what I ended up doing was introduce a simple boolean property named "useObjDomainAsKey" as part of the Query object.

I experimented with this being a property at the OutputWriter level and Query level and it seems to make a lot more sense at the Query level. 

If useObjDomainAsKey is set to true then JMXTrans will use the `Domain' of the MBean as part of the key instead of the class name (which is the default behavior). E.g. if 

ObjectName is: java.lang:type=Memory and ClassName is: sun.management.MemoryImpl then if useObjDomainAsKey is true the output will include java.lang instead of sun.management.MemoryImpl

It respects the allowDottedKeys property so the actual output could be either java_lang or java.lang

resultAlias will still overrule the domain type, just as it overrules the default ClassName behavior. 

I believe this would also solve this users request: https://groups.google.com/forum/#!topic/jmxtrans/Exhb4BEul-M

What do you think?

Conor
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/jmxtrans/jmxtrans/pull/264%23issuecomment-91378829%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/264%23issuecomment-91382635%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/264%23issuecomment-91382649%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/264%23discussion_r28111503%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/264%23discussion_r28112346%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/264%23discussion_r28113164%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/264%23issuecomment-91396311%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/264%23issuecomment-91439483%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/264%23discussion_r28205050%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/264%23discussion_r28205710%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/264%23issuecomment-91378829%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%27m%20not%20a%20huge%20fan%20of%20adding%20another%20configuration%20argument%2C%20but%20in%20general%2C%20this%20seems%20ok%20to%20me.%5Cr%5Cn%5Cr%5CnCould%20you%20remove%20the%20.settings%20folder%20and%20the%20.gitignore%20change%20in%20another%20PR%20though%3F%22%2C%20%22created_at%22%3A%20%222015-04-09T23%3A15%3A51Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/85355%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lookfirst%22%7D%7D%2C%20%7B%22body%22%3A%20%22Thanks.%20I%20committed%20again.%22%2C%20%22created_at%22%3A%20%222015-04-09T23%3A45%3A47Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/997809%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/conorbev%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-none%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222015-04-09T23%3A45%3A56Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/997809%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/conorbev%22%7D%7D%2C%20%7B%22body%22%3A%20%22Ok%2C%20just%20need%20to%20get%20%40gehel%27s%20opinion%20on%20this.%22%2C%20%22created_at%22%3A%20%222015-04-10T01%3A33%3A52Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/85355%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lookfirst%22%7D%7D%2C%20%7B%22body%22%3A%20%22Looks%20good%20to%20me%2C%20but%20I%27d%20like%20to%20have%20a%20deeper%20look%20this%20evening%20before%20merging.%22%2C%20%22created_at%22%3A%20%222015-04-10T05%3A55%3A36Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415765%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/gehel%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20d4c1513b88093e428af697e116b1c5d39009d205%20src/com/googlecode/jmxtrans/model/Query.java%2030%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/264%23discussion_r28205050%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%60isUse%60%20does%20not%20sound%20like%20valid%20english%20to%20me%20%28but%20English%20is%20not%20my%20first%20language%29.%20Wouldn%27t%20%60isUsingObjectDomainAsKey%60%20be%20more%20correct%3F%22%2C%20%22created_at%22%3A%20%222015-04-12T16%3A38%3A39Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415765%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/gehel%22%7D%7D%2C%20%7B%22body%22%3A%20%22You%20are%20right.%20%60isUsingObjDomainAsKey%60%20is%20better%20English.%5Cr%5Cn%5Cr%5CnHowever%20in%20the%20JSON%2C%20%60useObjDomainAsKey%60%20is%20more%20correct%2C%20it%20wouldn%27t%20sound%20valid%20to%20say%20%60using%60%20there%20%28also%20I%20shortened%20Object%20to%20Obj%20to%20match%20the%20existing%20%60obj%60%20parameter%20and%20make%20it%20clear%20that%20this%20property%20is%20referring%20to%20the%20same%20thing%29.%5Cr%5Cn%5Cr%5CnWhile%20%60isUseObjDomainAsKey%60%20isn%27t%20great%20English%20I%20thought%20it%20was%20clearer%20in%20the%20code%20to%20have%20the%20method%20name%20match%20the%20attribute%20it%20is%20returning%20exactly%2C%20and%20to%20follow%20the%20existing%20convention%20in%20the%20class%20to%20return%20boolean%20attributes%20using%20an%20%60is%60%20prefix%2C%20such%20as%20%60isAllowDottedKeys%60.%20I%20think%20it%27s%20a%20reasonable%20compromise%20and%20OK%20for%20it%20not%20to%20be%20perfect%20English%20in%20the%20code.%5Cr%5Cn%5Cr%5CnThat%20said%2C%20I%20don%27t%20have%20a%20strong%20opinion%20on%20it%20and%20am%20happy%20to%20change%20it%20if%20you%20prefer.%22%2C%20%22created_at%22%3A%20%222015-04-12T17%3A58%3A59Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/997809%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/conorbev%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20src/com/googlecode/jmxtrans/model/Query.java%3AL81-95%22%7D%2C%20%22Pull%202c48a93815b350fc7366beb88d3d6a62ee73de1c%20test/com/googlecode/jmxtrans/model/output/CloudWatchWriterTests.java%207%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/264%23discussion_r28111503%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Remove%20this.%20%3D%29%22%2C%20%22created_at%22%3A%20%222015-04-09T23%3A14%3A32Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/85355%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lookfirst%22%7D%7D%2C%20%7B%22body%22%3A%20%22Doh%2C%20thank%20you.%20I%20will%20clean%20up%20the%20.settings%20stuff%20too.%22%2C%20%22created_at%22%3A%20%222015-04-09T23%3A29%3A54Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/997809%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/conorbev%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-complete%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222015-04-09T23%3A45%3A57Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/997809%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/conorbev%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20test/com/googlecode/jmxtrans/model/output/CloudWatchWriterTests.java%3AL27-37%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/jmxtrans/jmxtrans/pull/264#issuecomment-91378829'>General Comment</a></b>
- <a href='https://github.com/lookfirst'></a> I'm not a huge fan of adding another configuration argument, but in general, this seems ok to me.
Could you remove the .settings folder and the .gitignore change in another PR though?
- <a href='https://github.com/conorbev'></a> Thanks. I committed again.
- <a href='https://github.com/conorbev'></a> :+1:<a href='#crh-update-none'></a>
- <a href='https://github.com/lookfirst'></a> Ok, just need to get @gehel's opinion on this.
- <a href='https://github.com/gehel'></a> Looks good to me, but I'd like to have a deeper look this evening before merging.
- [ ] <a href='#crh-comment-Pull d4c1513b88093e428af697e116b1c5d39009d205 src/com/googlecode/jmxtrans/model/Query.java 30'></a> &nbsp;<b><a href='https://github.com/jmxtrans/jmxtrans/pull/264#discussion_r28205050'>File: src/com/googlecode/jmxtrans/model/Query.java:L81-95</a></b>
- <a href='https://github.com/gehel'></a> `isUse` does not sound like valid english to me (but English is not my first language). Wouldn't `isUsingObjectDomainAsKey` be more correct?
- <a href='https://github.com/conorbev'></a> You are right. `isUsingObjDomainAsKey` is better English.
However in the JSON, `useObjDomainAsKey` is more correct, it wouldn't sound valid to say `using` there (also I shortened Object to Obj to match the existing `obj` parameter and make it clear that this property is referring to the same thing).
While `isUseObjDomainAsKey` isn't great English I thought it was clearer in the code to have the method name match the attribute it is returning exactly, and to follow the existing convention in the class to return boolean attributes using an `is` prefix, such as `isAllowDottedKeys`. I think it's a reasonable compromise and OK for it not to be perfect English in the code.
That said, I don't have a strong opinion on it and am happy to change it if you prefer.
- [x] <a href='#crh-comment-Pull 2c48a93815b350fc7366beb88d3d6a62ee73de1c test/com/googlecode/jmxtrans/model/output/CloudWatchWriterTests.java 7'></a> &nbsp;<b><a href='https://github.com/jmxtrans/jmxtrans/pull/264#discussion_r28111503'>File: test/com/googlecode/jmxtrans/model/output/CloudWatchWriterTests.java:L27-37</a></b>
- <a href='https://github.com/lookfirst'></a> Remove this. =)
- <a href='https://github.com/conorbev'></a> Doh, thank you. I will clean up the .settings stuff too.

<a href='https://www.codereviewhub.com/jmxtrans/jmxtrans/pull/264?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/jmxtrans/jmxtrans/pull/264?approve=1'></a>&nbsp;<a href='https://github.com/jmxtrans/jmxtrans/pull/264'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by gehel** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/jmxtrans/jmxtrans/pull/264'>Refresh</a>.<a href='#crh-update-none'></a>

File: test/com/googlecode/jmxtrans/model/output/CloudWatchWriterTests.java [hunk 1]
```diff
@@ -27,8 +27,10 @@ public void testValidationWithoutSettings() throws ValidationException {
 		Query query = Query.builder()
 				.setObj("test")
 				.build();
+		
 		Server server = Server.builder().setHost("localhost").setPort("123").build();
 		writer.validateSetup(server, query);
+		System.out.println("I got here");
```
Comment (lookfirst) [inline test/com/googlecode/jmxtrans/model/output/CloudWatchWriterTests.java]: Remove this. =)

**Comment by lookfirst** [comment]
I'm not a huge fan of adding another configuration argument, but in general, this seems ok to me.

Could you remove the .settings folder and the .gitignore change in another PR though?

File: test/com/googlecode/jmxtrans/model/output/CloudWatchWriterTests.java (hunk 1 above)
Comment (conorbev) [inline test/com/googlecode/jmxtrans/model/output/CloudWatchWriterTests.java]: Doh, thank you. I will clean up the .settings stuff too.

**Comment by conorbev** [comment]
Thanks. I committed again.

**Comment by conorbev** [comment]
:+1:<a href='#crh-update-none'></a>

File: test/com/googlecode/jmxtrans/model/output/CloudWatchWriterTests.java (hunk 1 above)
Comment (conorbev) [inline test/com/googlecode/jmxtrans/model/output/CloudWatchWriterTests.java]: :+1:<a href='#crh-update-complete'></a>

**Comment by lookfirst** [comment]
Ok, just need to get @gehel's opinion on this.

**Comment by gehel** [comment]
Looks good to me, but I'd like to have a deeper look this evening before merging.

File: src/com/googlecode/jmxtrans/model/Query.java [hunk 2]
```diff
@@ -78,6 +81,14 @@ public String getResultAlias() {
 	}
 
 	/**
+	 * The useObjDomainAsKey property allows you to specify the use of the Domain portion of the Object Name
+	 * as part of the output key instead of using the ClassName of the MBean which is the default behavior.
+	 */
+	public boolean isUseObjDomainAsKey() {
```
Comment (gehel) [inline src/com/googlecode/jmxtrans/model/Query.java]: `isUse` does not sound like valid english to me (but English is not my first language). Wouldn't `isUsingObjectDomainAsKey` be more correct?

Comment (conorbev) [inline src/com/googlecode/jmxtrans/model/Query.java]: You are right. `isUsingObjDomainAsKey` is better English.

However in the JSON, `useObjDomainAsKey` is more correct, it wouldn't sound valid to say `using` there (also I shortened Object to Obj to match the existing `obj` parameter and make it clear that this property is referring to the same thing).

While `isUseObjDomainAsKey` isn't great English I thought it was clearer in the code to have the method name match the attribute it is returning exactly, and to follow the existing convention in the class to return boolean attributes using an `is` prefix, such as `isAllowDottedKeys`. I think it's a reasonable compromise and OK for it not to be perfect English in the code.

That said, I don't have a strong opinion on it and am happy to change it if you prefer.

**Comment by gehel** [comment]
That's good enough for me at this point. There is some cleanup needed on how configuration parameters are passed around anyway ...

Thanks for your help!

File: src/com/googlecode/jmxtrans/model/Query.java (hunk 2 above)
Comment (lookfirst) [inline src/com/googlecode/jmxtrans/model/Query.java]: I posit that naming things is one of the hardest parts of software engineering. I routinely get it wrong. :cry:

Comment (gehel) [inline src/com/googlecode/jmxtrans/model/Query.java]: Naming and cache invalidation ...

**Pull request merged**